target/linux/lantiq/xrx200/base-files/etc/hotplug.d/firmware/11-ath10k-caldata

```diff
@@ -5,9 +5,16 @@
 . /lib/functions/caldata.sh
 
 case "$FIRMWARE" in
+"ath10k/cal-pci-0000:00:00.0.bin"|\
 "ath10k/cal-pci-0000:02:00.0.bin")
 	board=$(board_name)
 	case $board in
+		avm,fritz3490|avm,fritz3490-micron|avm,fritz5490|avm,fritz5490-micron|avm,fritz7490|avm,fritz7490-micron)
+			[ ! -d /lib/firmware/ath10k ] && mkdir /lib/firmware/ath10k
+			offset=$(hexdump -v -e '1/1 "%02x"' $(find_mtd_chardev "urlader") | awk '{print index($0, "4408")}')
```

Comment:
Can you explain what the shell hackery does/is supposed to do?

Reply:
@mkresin During testing there were reports that same devices do not even have the same offset of caldata for ath10k. So someone came up with the x'4408' indicator for the start. So this script searches for that hex value, calculates the offset. So it works for all devices so far without having to specify a fixed offset.

Reply:
What I can say is, that there is some kind of structure in the urlader for data like the wireless calibration. No idea if it's a table of contents or just magic values which mark the start of the caldata.

Using a fixed offset for the caldata as we are using it for all AVM devices is fragile, if not wrong, as you already noticed.

Would be better to understand how things are organised in the urlader.

Reply:
In a6f85b81b7efd3fa3265d70e5406275b2dc03f5f we just try to extract the cal_data from multiple offsets and hope the correct is one of them.

Reply:
@hauke This only works, because fritz_cal_extract is a binary and sets a return code, caldata_extract function from lib/functions/caldata.sh has no return code. Or should I move the hex hackery to a new fritz_ath10k_cal_extract utility in fritz-tools?

Reply:
I tried to research the urlader structure, but I can only get to the key value pairs. There is a key for the caldata offset, but in the fritzbox devices I have, its not set, so cannot be used. I have not found a way or an offset to start from to reliably get the the start offset of the caldata.

Reply:
The offset to read the ath9k caldata seems to be the same for all devices 0x1541 with length 0x440 and reverse extract (so ath9k caldata ends at 0x1981, which is actually the first byte). For my 7490 the ath10k caldata starts at 0x199F (0x1541 + 0x440 + 1E). In between the two caldata the first bytes are 0x5f 0x31 0x4C 0x41 giving no clue where caldata starts.
0x1541, 0x1981, 0x199F or 0x21E3 are not in the urlader, not even with swapped bytes.
I wrote a small program that starts at 0x199F going backwards to check if there is any two byte field between offset 0 and 0x199F that points as an offset to 0x199F.
The urlader variable area at 0x580 only points to all the key value pairs and points at the end (0x096A). After that there are 28 bytes before the 0x01FF 0x5AA5 id that I guess indicates the end of the key value section.
Within those bytes at offset 0x98A there is value 0x1802 that adds to 0x218C which is pretty close to the end of the ath10k caldata (0x199F + 0x844) 0x21E3, but misses 87 bytes. After 0x21E3 there is a lot of 0xFF following.
Between 0x1541 and 0x199F there are 28 bytes plus 0x01FF (which in tffs is also used as the end identifier).

I see no other chance than searching for 0x4408 for the ath10k caldata start.

Reply:
ok fine with me. 

Could you add this to the commit comment please so someone could fine this explanation when he looks at this code in 5 years.